WordPress/Classes/Utility/ContextManager.m

```diff
@@ -133,9 +133,32 @@ - (void)saveDerivedContext:(NSManagedObjectContext *)context withCompletionBlock
 
 - (void)saveContextAndWait:(NSManagedObjectContext *)context
 {
-    [context performBlockAndWait:^{
-        [self internalSaveContext:context];
-    }];
+    if (context != self.mainContext) {
+        [context performBlockAndWait:^{
+            NSError *error = nil;
+            
+            if ([context hasChanges] && ![context save:&error]) {
+                DDLogError(@"Fatal Core Data Error encountered — throwing an exception. Underlying error is:\n %@", error);
+                @throw [NSException exceptionWithName:@"Unresolved Core Data save error"
+                                               reason:[NSString stringWithFormat:@"Unresolved Core Data save error - derived context. Core Data Error Domain: %@, code: %i", error.domain, error.code]
+                                             userInfo:error.userInfo];
+            }
+        }];
+        
+        if (context.parentContext != nil) {
+            // It's not ideal for a context to be the one to define when its parent is saved, but in order
+            // to follow the existing architecture, we can at least ensure saving the parent is safe, and
+            // that we're not deadlocking the current thread to do that.
+            //
+            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
```

Comment:
I'm not sure I understand. The way I understood the problem is that, as you described in https://github.com/wordpress-mobile/WordPress-iOS/pull/12459/files#r321340875, we want the data to be persisted right away so that any fetch executing somewhere else would receive the up to date data. But doesn't running `dispatch_async` here prevent the `context.parentContext` to be persisted **right away**? And hence, it's still possible for fetch requests somewhere else to receive stale data?

What was causing the deadlock in here anyway?

Reply:
> But doesn't running dispatch_async here prevent the context.parentContext to be persisted right away?

This is correct, but we only need it to be persisted right away to the main context - saving our derived context (the context the service is using) does just that.  Once the changes hit the main context, they're available across the entire App.

The save to disk (which is what happens when we save the parent context / main context) can be asynchronous (in fact due to our `ContextManager`'s architecture [it's always async no matter what we do](https://github.com/wordpress-mobile/WordPress-iOS/blob/b862ae4ad4f50748bb02b98bdab694f4d197da22/WordPress/Classes/Utility/ContextManager.m#L277-L290)).

As a matter of fact I don't think a derived context should attempt to save to disk at all, or to save parent contexts but that's a deeper discussion beyond the scope of this PR.  This is why I don't feel it'd be right to look for deeper changes in this PR.

> And hence, it's still possible for fetch requests somewhere else to receive stale data?

Not with the code as it is.

This could only happen if we called `uploadMedia()` with a context that doesn't have the main context as it's parent.  But again, picking a context that has the main context as its parent is a deliberate choice and the responsibility of the caller.  This part is important - not any context is the same when calling a service.